crypto/bn/asm/rsaz-2k-avx512.pl

```diff
@@ -546,36 +542,36 @@ sub amm52x20_x1_norm {
     vpbroadcastq    $red_tbl_idx, $idx
     leaq   `(1<<5)*2*20*8`($red_tbl), %rax  # holds end of the tbl
 
-    vpxor   $t4xmm, $t4xmm, $t4xmm
-    vmovdqa64   $t4, $t3                    # zeroing t0..4, cur_idx
-    vmovdqa64   $t4, $t2
-    vmovdqa64   $t4, $t1
-    vmovdqa64   $t4, $t0
-    vmovdqa64   $t4, $cur_idx
+    vpxor   $t0xmm, $t0xmm, $t0xmm
+    vmovdqa64   $t0, $t4                    # zeroing t0..4, cur_idx
+    vmovdqa64   $t0, $t3
+    vmovdqa64   $t0, $t2
+    vmovdqa64   $t0, $t1
+    vmovdqa64   $t0, $cur_idx
 
 .align 32
 .Lloop:
-    vpcmpq  \$0, $cur_idx, $idx, %k1        # mask of (idx == cur_idx)
-    addq    \$320, $red_tbl                 # 320 = 2 * 20 digits * 8 bytes
-    vpaddq  $ones, $cur_idx, $cur_idx       # increment cur_idx
-    vmovdqu64  -320($red_tbl), $tmp0        # load data from red_tbl
-    vmovdqu64  -288($red_tbl), $tmp1
-    vmovdqu64  -256($red_tbl), $tmp2
-    vmovdqu64  -224($red_tbl), $tmp3
-    vmovdqu64  -192($red_tbl), $tmp4
-    vpblendmq  $tmp0, $t0, ${t0}{%k1}       # extract data when mask is not zero
+    vpcmpq  \$0, $cur_idx, $idx, %k1           # mask of (idx == cur_idx)
+    addq    \$`2*20*8`, $red_tbl
+    vpaddq  $ones, $cur_idx, $cur_idx          # increment cur_idx
+    vmovdqu64  `-2*20*8+0*32`($red_tbl), $tmp0 # load data from red_tbl
+    vmovdqu64  `-2*20*8+1*32`($red_tbl), $tmp1
+    vmovdqu64  `-2*20*8+2*32`($red_tbl), $tmp2
+    vmovdqu64  `-2*20*8+3*32`($red_tbl), $tmp3
+    vmovdqu64  `-2*20*8+4*32`($red_tbl), $tmp4
+    vpblendmq  $tmp0, $t0, ${t0}{%k1}          # extract data when mask is not zero
     vpblendmq  $tmp1, $t1, ${t1}{%k1}
     vpblendmq  $tmp2, $t2, ${t2}{%k1}
     vpblendmq  $tmp3, $t3, ${t3}{%k1}
     vpblendmq  $tmp4, $t4, ${t4}{%k1}
```

Comment:
how about using vmovdqu64 with writemask option (as per _mm256_mask_loadu_epi64() intrinsic) - with this it looks to be possible to achieve the same result as combining vmovdqu64 and vpblendmq. Then the 5 X vpblendmq could be removed.
One reason could be that if the mask is all zeros, then there may be no load from memory which is required for constant time. But if memory is guaranteed to be loaded, then it could be an option to investigate.

Reply:
It was done intentionally. We cannot be sure that load from memory with mask will be executed constant time. That is why we have here unconditional load and then reg-reg mask operation which is ok.

Reply:
makes sense, thanks